utilities/transactions/pessimistic_transaction.cc

```diff
@@ -135,6 +136,486 @@ WriteCommittedTxn::WriteCommittedTxn(TransactionDB* txn_db,
                                      const TransactionOptions& txn_options)
     : PessimisticTransaction(txn_db, write_options, txn_options) {}
 
+Status WriteCommittedTxn::GetForUpdate(const ReadOptions& read_options,
+                                       ColumnFamilyHandle* column_family,
+                                       const Slice& key, std::string* value,
+                                       bool exclusive, const bool do_validate) {
+  column_family =
+      column_family ? column_family : db_impl_->DefaultColumnFamily();
+  assert(column_family);
+  if (!read_options.timestamp) {
+    const Comparator* const ucmp = column_family->GetComparator();
+    assert(ucmp);
+    size_t ts_sz = ucmp->timestamp_size();
+    if (0 == ts_sz) {
+      return TransactionBaseImpl::GetForUpdate(read_options, column_family, key,
+                                               value, exclusive, do_validate);
+    }
+  } else {
+    Status s = db_impl_->FailIfTsSizesMismatch(column_family,
+                                               *(read_options.timestamp));
+    if (!s.ok()) {
+      return s;
+    }
+  }
+
+  if (kMaxTxnTimestamp == read_timestamp_) {
```

Comment:
I may be missing something, but we do perform check for `do_validate`. Actually, we require `do_validate` to be true, otherwise, it's an invalid argument.

Reply:
I'm talking about the case when `read_timestamp_` is not set (`kMaxTxnTimestamp`). The current logic considers that an error even if `do_validate` is false, right? I'm wondering if that's what we want, or if we would want to make sure `read_timestamp_` is set _if and only if_ `do_validate` is set.

Reply:
If the control flow reaches here, it means we enable timestamp and call GetForUpdate(). I think in this case, we should just require do_validate to be true AND read_timestamp be set.

Reply:
Sorry, I don't follow... At the point where we check `kMaxTxnTimestamp == read_timestamp_`, is `do_validate` guaranteed to be set? If it is, the `else if (!do_validate)` branch can't be hit, right?

In other words, I was wondering if we would want to do some version of this:

```
if (do_validate) {
   if (kMaxTxnTimestamp == read_timestamp_) {
      // error
   }
} else {
   if (kMaxTxnTimestamp != read_timestamp_) {
      // error
   }
}
```

Reply:
I may know the confusion. I think the logic can be
```
if (!do_validate) {
  // error. Because GetForUpate() with timestamp enabled must perform validation.
} else if (kMaxTxnTimestamp == read_timestamp_) {
  // error
}
```

Reply:
Thanks! Checking `do_validate` first definitely makes the intent clearer to me